adapters/sparteo/sparteo.go

```diff
@@ -162,6 +190,148 @@ func (a *adapter) MakeRequests(req *openrtb2.BidRequest, reqInfo *adapters.Extra
 	return []*adapters.RequestData{requestData}, errs
 }
 
+func normalizeHostname(host string) string {
+	host = strings.TrimSpace(host)
+	if host == "" {
+		return ""
+	}
+
+	u, err := url.Parse(host)
+	if err != nil || u.Hostname() == "" {
+		if i := strings.Index(host, ":"); i >= 0 {
+			host = host[:i]
+		} else if i := strings.Index(host, "/"); i >= 0 {
+			host = host[:i]
+		}
+	} else {
+		host = u.Hostname()
+	}
+
+	host = strings.ToLower(host)
+	host = strings.TrimSuffix(host, ".")
+	host = strings.TrimPrefix(host, "www.")
+
+	if host == "null" {
+		return ""
+	}
+	return host
+}
+
+func resolveSiteDomain(site *openrtb2.Site) *string {
+	if site != nil {
+		if d := normalizeHostname(site.Domain); d != "" {
+			return ptrutil.ToPtr(d)
+		}
+		if fromPage := normalizeHostname(site.Page); fromPage != "" {
+			return ptrutil.ToPtr(fromPage)
+		}
+		return ptrutil.ToPtr(unknownValue)
+	}
+	return nil
+}
+
+func resolveAppDomain(app *openrtb2.App) *string {
+	if app != nil {
+		if d := normalizeHostname(app.Domain); d != "" {
+			return ptrutil.ToPtr(d)
+		}
+		return ptrutil.ToPtr(unknownValue)
+	}
+	return nil
+}
+
+func resolveBundle(app *openrtb2.App) *string {
+	if app == nil {
+		return nil
+	}
+
+	raw := app.Bundle
+	if strings.TrimSpace(raw) == "" {
+		return ptrutil.ToPtr(unknownValue)
+	}
+
+	b := strings.TrimSpace(raw)
+	if strings.EqualFold(b, "null") {
+		return ptrutil.ToPtr(unknownValue)
+	}
+
+	return ptrutil.ToPtr(b)
+}
+
+func ensurePublisher(p *openrtb2.Publisher) *openrtb2.Publisher {
+	if p == nil {
+		p = &openrtb2.Publisher{}
+	}
+	if p.Ext == nil {
+		p.Ext = jsonutil.RawMessage("{}")
+	}
+
+	return p
+}
+
+func updatePublisherExtension(targetExt *jsonutil.RawMessage, networkID, fieldPath string) ([]byte, error) {
+	var pubExt map[string]interface{}
+	if err := jsonutil.Unmarshal(*targetExt, &pubExt); err != nil {
+		pubExt = make(map[string]interface{})
+	}
+
+	params, ok := pubExt["params"].(map[string]interface{})
+	if !ok {
+		params = make(map[string]interface{})
+		pubExt["params"] = params
+	}
+	params["networkId"] = networkID
+
+	updated, err := jsonutil.Marshal(pubExt)
+	if err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("Error marshaling %s: %s", fieldPath, err),
+		}
+	}
+	return updated, nil
+}
+
+func (a *adapter) buildEndpointURL(networkId string, siteDomain string, appDomain string, bundle string) (string, error) {
```

Comment:
Why dont you build the url via the templated endpoint, if any param isn't present use the unknown value. You have a lot of code that could be simplified using this approach

Example here: https://github.com/prebid/prebid-server/blob/master/adapters/adview/adview.go#L163